src/lib/common/sol-platform-linux-common.c

```diff
@@ -263,3 +307,581 @@ sol_platform_impl_get_os_version(char **version)
+            uevent.modalias = sol_str_slice_from_str(++del);
+        else if (!strncmp(curr, "ACTION", value_len))
+            uevent.action = sol_str_slice_from_str(++del);
+        else if (!strncmp(curr, "SUBSYSTEM", value_len))
+            uevent.subsystem = sol_str_slice_from_str(++del);
+        else if (!strncmp(curr, "DEVTYPE", value_len))
+            uevent.devtype = sol_str_slice_from_str(++del);
+        else if (!strncmp(curr, "DEVNAME", value_len))
+            uevent.devname = sol_str_slice_from_str(++del);
+
+        i += strlen(msg + i) + 1;
+    }
+
+    sol_uevent_event_dispatch(ctx, &uevent);
+}
+
+static bool
+sol_uevent_handler(void *data, int fd, unsigned int cond)
+{
+    int len;
+    struct uevent_context *ctx = data;
+    char buffer[512];
+
+    len = recv(ctx->uevent.fd, buffer, sizeof(buffer), MSG_WAITALL);
+    if (len == -1) {
+        SOL_ERR("Could not read netlink socket. %s", sol_util_strerrora(errno));
+        return false;
+    } else {
+        sol_uevent_read_msg(ctx, buffer, len);
+    }
+
+    return true;
+}
+
+static int
+sol_uevent_register(struct uevent_context *ctx)
+{
+    struct sockaddr_nl nls;
+
+    memset(&nls, 0, sizeof(struct sockaddr_nl));
+    nls.nl_family = AF_NETLINK;
+    nls.nl_pid = getpid();
+    nls.nl_groups = -1;
+
+    ctx->uevent.fd = socket(PF_NETLINK, SOCK_DGRAM | SOCK_CLOEXEC,
+        NETLINK_KOBJECT_UEVENT);
+    if (ctx->uevent.fd == -1) {
+        SOL_ERR("Could not open uevent netlink socket.");
+        return -errno;
+    }
+
+    if (bind(ctx->uevent.fd, (void *)&nls, sizeof(struct sockaddr_nl))) {
+        SOL_ERR("Could not bind to uevent socket.");
+        close(ctx->uevent.fd);
+        ctx->uevent.fd = 0;
+        return -errno;
+    }
+
+    ctx->uevent.watch = sol_fd_add(ctx->uevent.fd, SOL_FD_FLAGS_IN,
+        sol_uevent_handler, ctx);
+    return 0;
+}
+
+SOL_API int
+sol_platform_linux_uevent_subscribe(const char *action, const char *subsystem,
+    void (*uevent_cb)(void *cb_data, struct sol_uevent *uevent),
+    const void *cb_data)
+{
+    struct uevent_callback *cb;
+
+    if (!uevent_ctx.running)
+        sol_vector_init(&uevent_ctx.callbacks, sizeof(struct uevent_callback));
+
+    cb = sol_vector_append(&uevent_ctx.callbacks);
+    SOL_NULL_CHECK(cb, -1);
+
+    cb->action = action;
+    cb->subsystem = subsystem;
+    cb->uevent_cb = uevent_cb;
+    cb->cb_data = cb_data;
+
+    if (!uevent_ctx.running) {
+        sol_uevent_register(&uevent_ctx);
+    }
+
+    uevent_ctx.running = true;
+    return 0;
+}
+
+static void
+sol_uevent_cleaup(struct uevent_context *ctx)
+{
+    if (ctx->uevent.watch)
+        sol_fd_del(ctx->uevent.watch);
+
+    if (ctx->uevent.fd)
+        close(ctx->uevent.fd);
+
+    sol_vector_clear(&ctx->callbacks);
+    ctx->running = false;
+}
+
+SOL_API int
+sol_platform_linux_uevent_unsubscribe(const char *action, const char *subsystem,
+    void (*uevent_cb)(void *cb_data, struct sol_uevent *uevent))
```

Comment:
you're not using `action` and `subsystem`. And always provide the `data` to verify the callback you're using.

When people program using wrappers, like doing a Language binding (Python/JS) or just helpers, he will give you the same callback function for all cases, what changes is the context (void *data).